test/e2e-monitoring/monitoring_test.go

```diff
@@ -0,0 +1,125 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/dynamic"
+)
+
+func TestResourceVersionApplication(t *testing.T) {
+	config, err := library.NewClientConfigForTest()
+	require.NoError(t, err)
+
+	// Define the resource.
+	duration := monv1.Duration("5m")
+	resource := monv1.PrometheusRule{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "monitoring.coreos.com/v1",
+			Kind:       "PrometheusRule",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-prometheusrule",
+			Namespace: "default",
+		},
+		Spec: monv1.PrometheusRuleSpec{
+			Groups: []monv1.RuleGroup{
+				{
+					Name: "example",
+					Rules: []monv1.Rule{
+						{
+							Alert: "Foo",
+							Expr:  intstr.FromString("foo > 0"),
+							For:   &duration,
+							Labels: map[string]string{
+								"bar": "baz",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	gvr := schema.GroupVersionResource{
+		Group:    resource.GroupVersionKind().Group,
+		Version:  resource.GroupVersionKind().Version,
+		Resource: "prometheusrules",
+	}
+
+	// Initialize pre-requisites.
+	cache := resourceapply.NewResourceCache()
+	recorder := events.NewInMemoryRecorder("TestResourceVersionApplication")
+	dynamicClient, err := dynamic.NewForConfig(config)
+	require.NoError(t, err)
+
+	// Create the resource.
+	unstructuredResource := &unstructured.Unstructured{}
+	unstructuredResourceMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&resource)
+	require.NoError(t, err)
+	unstructuredResource.SetUnstructuredContent(unstructuredResourceMap)
+	_, didUpdate, err := resourceapply.ApplyUnstructuredResourceImproved(
+		context.TODO(),
+		dynamicClient,
+		recorder,
+		unstructuredResource,
+		cache,
+		gvr,
+		nil,
+		nil,
+	)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		t.Fatalf("Failed to create resource: %v", err)
+	}
+	require.True(t, didUpdate)
+
+	// Update the resource with a change, without specifying a resource version.
+	resource.Spec.Groups[0].Rules[0].Labels["bar"] = "qux"
+	unstructuredResourceMap, err = runtime.DefaultUnstructuredConverter.ToUnstructured(&resource)
+	require.NoError(t, err)
+	unstructuredResource.SetUnstructuredContent(unstructuredResourceMap)
+	_, didUpdate, err = resourceapply.ApplyUnstructuredResourceImproved(
+		context.TODO(),
+		dynamicClient,
+		recorder,
+		unstructuredResource,
+		cache,
+		gvr,
+		nil,
+		nil,
+	)
+	if err != nil {
+		t.Fatalf("Failed to update resource: %v", err)
+	}
+	require.True(t, didUpdate)
+
+	// Update the resource without any changes, without specifying a resource version.
+	_, didUpdate, err = resourceapply.ApplyUnstructuredResourceImproved(
+		context.TODO(),
+		dynamicClient,
+		recorder,
+		unstructuredResource,
+		cache,
+		gvr,
+		nil,
+		nil,
+	)
+	if err != nil {
+		t.Fatalf("Failed to update resource: %v", err)
+	}
+	require.False(t, didUpdate)
+
```

Comment:
Would it make sense to GET the updated obj and compare it to an expected one?

Reply:
Added, PTAL.